pkg/assetfetcher/http.go

```diff
@@ -0,0 +1,162 @@
+		errChan <- &NexposeHTTPRequestError{err, req.URL.String()}
+		return nil, errChan
+	}
+	defer res.Body.Close()
+	respBody, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		errChan <- &ErrorReadingNexposeResponse{err, req.URL.String()}
+		return nil, errChan
+	}
+
+	var siteAssetResp SiteAssetsResponse
+	if err := json.Unmarshal(respBody, &siteAssetResp); err != nil {
+		errChan <- &ErrorParsingJSONResponse{err, req.URL.String()}
+		return nil, errChan
+	}
+
+	pagedAssetChan := make(chan domain.Asset, siteAssetResp.Page.TotalResources)
+	pagedErrChan := make(chan error, siteAssetResp.Page.TotalResources)
+
+	for _, asset := range siteAssetResp.Resources {
+		pagedAssetChan <- asset
+	}
+
+	// We've gotten the first page (page 0) and added the assets to the channel,
+	// so here we'll get the TotalPages of assets that Nexpose has, then paginate
+	// through from page 1 to TotalPages to get all the assets
+	totalPages := siteAssetResp.Page.TotalPages
+	var wg sync.WaitGroup
+	for currentPage := 1; currentPage < totalPages; currentPage++ {
+		wg.Add(1)
+		go c.makeRequest(ctx, &wg, siteID, currentPage, pagedAssetChan, pagedErrChan)
+	}
+
+	go func() {
+		defer close(pagedErrChan)
+		defer close(pagedAssetChan)
+		wg.Wait()
+	}()
+
+	return pagedAssetChan, pagedErrChan
+}
+
+func (c *NexposeAssetFetcher) makeRequest(ctx context.Context, wg *sync.WaitGroup, siteID string, page int, assetChan chan domain.Asset, errChan chan error) {
+	defer wg.Done()
+	req, err := newNexposeSiteAssetsRequest(c.Host, siteID, page, c.PageSize)
+	if err != nil {
+		errChan <- err
+		return
+	}
+	res, err := c.HTTPClient.Do(req.WithContext(ctx))
+	if err != nil {
+		errChan <- &NexposeHTTPRequestError{err, req.URL.String()}
+		return
+	}
+	defer res.Body.Close()
+	respBody, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		errChan <- &ErrorReadingNexposeResponse{err, req.URL.String()}
+		return
+	}
+	var siteAssetResp SiteAssetsResponse
+	if err := json.Unmarshal(respBody, &siteAssetResp); err != nil {
+		errChan <- &ErrorParsingJSONResponse{err, req.URL.String()}
+		return
+	}
+	for _, asset := range siteAssetResp.Resources {
+		assetChan <- asset
+	}
+}
+
+// newNexposeSiteAssetsRequest builds URL we'll use to make the request to Nexpose's site assets endpoint
+func newNexposeSiteAssetsRequest(host string, siteID string, page int, size int) (*http.Request, error) {
+	u, err := url.Parse(host)
+	if err != nil {
+		return nil, &URLParsingError{err, host}
+	}
+	u.Path = path.Join(u.Path, basePath, nexposeSite, siteID, nexposeAssets)
+	q := u.Query()
+	q.Set(pageQueryParam, fmt.Sprint(page))
+	q.Set(sizeQueryParam, fmt.Sprint(size))
+	u.RawQuery = q.Encode()
+	req, _ := http.NewRequest(http.MethodGet, u.String(), http.NoBody)
+	// the only time http.NewRequest returns an error is if there's a parsing error,
+	// which we already checked for earlier, so no need to check it again
+	req.SetBasicAuth(os.Getenv("NEXPOSE_USERNAME"), os.Getenv("NEXPOSE_PASSWORD"))
```

Comment:
Typically these would be added as Attributes to the struct. It would also make your refactoring a little easier once you start the work to use the settings project.

Reply:
Agree, but I really need to stop iterating and making changes to this issue and PR. I'll add this work to the issue where I incorporate the settings project. Thanks for the heads up on this